p2panda-rs/src/storage_provider/models.rs

```diff
@@ -0,0 +1,99 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+//! Data models which are used in StorageProvider.
+use std::fmt::Debug;
+
+use crate::document::DocumentId;
+use crate::entry::{decode_entry, EntrySigned, LogId};
+use crate::identity::Author;
+use crate::operation::OperationEncoded;
+use crate::schema::SchemaId;
+use crate::storage_provider::ValidationError;
+use crate::Validate;
+
+/// Struct wrapping an entry with it's operation.
+///
+/// Used internally throughout `storage_provider` in method args and default trait definitions.
+/// The `AsStorageEntry` trait requires `TryFrom<EntryWithOperation>` & `TryInto<EntryWithOperation>`
+/// conversion traits to be present.
+#[derive(Debug, Clone)]
+pub struct EntryWithOperation(EntrySigned, OperationEncoded);
+
+impl EntryWithOperation {
+    /// Instantiate a new EntryWithOperation.
+    pub fn new(entry: &EntrySigned, operation: &OperationEncoded) -> Result<Self, ValidationError> {
+        let entry_with_operation = Self(entry.clone(), operation.clone());
+        entry_with_operation.validate()?;
+        Ok(entry_with_operation)
+    }
+
+    /// Returns a reference to the encoded entry.
+    pub fn entry_encoded(&self) -> &EntrySigned {
```

Comment:
Maybe we call this `entry_signed`? It is also encoded but knowing the p2panda API (maybe too much :sweat_smile:), I expected it to have the same name as the struct.